### ISSUE TITLE
in_gecko feature additive?

### ISSUE THREAD

**Issue by Dushistov** [rusqlite/rusqlite#750]
As mention on #747 I am working on wrapper for C++ (static) library that include sqlite3.o.
So I need to tell rusqlite that I already has sqlite3.o and `in_gecko` comes to play.
Let's suppose I have workspace with such crates:
**a** depend on rusqlite
**b** depend on C++ library that include sqlite3.o
**main** depend on **a** and **b**, so
```
rusqlite = { version = "0.23.1", features = ["in_gecko"] }
```
and I can build and run **main** without problem.

But then I run `cargo test --all` in the root of workspace and all failed.
Because of cargo thinks that all features are additive,
it just build all crates with rusqlite features = "in_gecko" ,
but **a** have no dependency from "C++ library" and build of tests failed.

I can not add dependency to **a**, because of it also used by **main II**, which should not include "C++ library".

**Comment by thomcc** [comment]
Right, there's not really a solution here sadly. This is a limitation of cargo features, and even if the in_gecko feature were made into less of a hack, it would still exist.

I don't really have advice here, unfortunately.

**Comment by Dushistov** [comment]
Ideally in_gecko should not exists in configuration, it should be calculated automatically.
Like in `build.rs`  of my `c++-library-crate-sys` invokes:
```
println!("cargo:I_HAVE_SQLITE");
```
and then `build.rs` from `libsqlite3-sys` detect this flag and act like `in_gecko` is settted.

But, I am not sure here, correct me please,
currently  `libsqlite3-sys` should depend on `c++-library-crate-sys` to make it possible?
While I need other way around pass from **main** to **libsqlite3-sys**,
and that is impossible: rust-lang/cargo#4121

Another option is use environment variable, and run to build executable
```
IN_GECKO=1 cargo build -p main
```
that almost works, main executable is build via wrapper around `cargo`,
but I need special case for testing **b**: `IN_GECKO=1 cargo test -p b`,
instead of `cargo test --all`.

What do you think about environment variable in addition to `in_gecko`?

**Comment by thomcc** [comment]
I'd accept a patch for an env variable. In addition to in_gecko. Don't call it IN_GECKO though. Calling it `LIBSQLITE3_SYS_NO_LINK` would be more accurate.

**Comment by mhammond** [comment]
I'm playing a little with `in_gecko` (thanks @thomcc!) because we are toying with the possibility of having our rusqlite consuming crates use the version of sqlite on iOS. It turns out the `in_gecko` feature works in that scenario, which shouldn't be a surprise, but does highlight the name isn't quite correct.

I was thinking about this issue and ISTM that the current `in_gecko` might be able to be implied by the *lack of* the bundled feature. This causes more work for some consumers who need this (specifically, they probably can not specify this feature as a `[dependency]`, but *would* need to include it as a `[dev-dependency]` so the final test executables can link.

This *almost* works correctly, except for how `build.rs` operates when the bundled feature is *not* specified - ie, [this branch](https://github.com/rusqlite/rusqlite/blob/c2fbd167dec8b043524a11df8a471120008669dd/libsqlite3-sys/build.rs#L84) looks almost identical to what happens when `bundled` is enabled. However, if it is replaces with the action taken for `in_gecko`, then things work as I expect (at least in my fairly limited testing)

If that was possible, I think it would largely close this issue and end up with the features being purely additive. It probably wouldn't impact most consumers - they could continue to specify bundled as a feature in their main dependencies so long as they were never built in an environment requiring this capability.

But what I don't understand is what that change would break - ie, why that default branch linked above does what it does. I can't see any reference in the docs either. Can someone please enlighten me?
